log, sync: change lock contention from preprocessor directive to log category

jonatack opened the pull request:
To enable lock contention logging, `DEBUG_LOCKCONTENTION` has to be defined at compilation. Once built, the logging is not limited to a category and is high frequency, verbose and in all-caps. With these factors combined, it seems likely to be rarely used.

This patch:
- adds a `lock` logging category
- adds a timing macro in microseconds, `LOG_TIME_MICROS_WITH_CATEGORY`
- updates `BCLog::LogMsg()` to omit irrelevant decimals for microseconds and skip unneeded code and math
- improves the lock contention logging, drops the all-caps, and displays the duration in microseconds
- removes the conditional compilation directives
- allows lock contentions to be logged on startup with `-debug=lock` or at run time with `bitcoin-cli logging '["lock"]'`

```
$ bitcoind -signet -debug=lock
2021-09-01T12:40:01Z LockContention: cs_vNodes, net.cpp:1920 started
2021-09-01T12:40:01Z LockContention: cs_vNodes, net.cpp:1920 completed (4μs)
2021-09-01T12:40:01Z LockContention: cs_vNodes, net.cpp:1302 started
2021-09-01T12:40:01Z LockContention: cs_vNodes, net.cpp:1302 completed (4μs)
2021-09-01T12:40:02Z LockContention: cs_vNodes, net.cpp:2242 started
2021-09-01T12:40:02Z LockContention: cs_vNodes, net.cpp:2242 completed (20μs)
2021-09-01T12:43:04Z LockContention: ::cs_main, validation.cpp:4980 started
2021-09-01T12:43:04Z LockContention: ::cs_main, validation.cpp:4980 completed (3μs)

$ bitcoin-cli -signet logging
  "lock": true,

$ bitcoin-cli -signet logging [] '["lock"]'
  "lock": false,

$ bitcoin-cli -signet logging '["lock"]'
  "lock": true,
```

I've tested this with Clang 13 and GCC 10.2.1, on Debian, with and without `--enable-debug`.

jonatack committed:
log, sync: add LOCK logging category, apply it to lock contention

jonatack committed:
sync: remove DEBUG_LOCKCONTENTION preprocessor directives

to allow logging the lock contentions without the need to define
DEBUG_LOCKCONTENTION at compile time.

jonatack commented:
Seeing a CI failure; converting to draft status. 
```
sync.cpp:27:1: error: static_assert failed "thread_local is not supported"
static_assert(false, "thread_local is not supported");
^             ~~~~~
1 error generated.
```

laanwj commented on src/sync.h:
```diff
@@ -138,14 +137,10 @@ class SCOPED_LOCKABLE UniqueLock : public Base
     {
         EnterCritical(pszName, pszFile, nLine, Base::mutex());
-#ifdef DEBUG_LOCKCONTENTION
         if (!Base::try_lock()) {
```
I am somewhat worried that the extra calls to `Base::try_lock` cause performance overhead in the non-lock contention case. We might want to gate *that too* on the log category being enabled.

jonatack replied:
Thanks, wondering that as well (or alternatively, leave the directives in place and just improve the logging experience when `DEBUG_LOCKCONTENTION` is defined).

jonatack replied:
Implemented your suggestion (thanks!)

jonatack committed:
log, sync: improve lock contention logging and add time duration

in microseconds.

Change the function name in order to print "LockContention" instead
of "PrintLockContention" to the log.  Add Doxygen documentation.

With this change, the lock contention log prints:

2021-09-01T11:29:03Z LockContention: pnode->cs_vSend, net.cpp:1373 started
2021-09-01T11:29:03Z LockContention: pnode->cs_vSend, net.cpp:1373 completed (31μs)
2021-09-01T11:29:03Z LockContention: cs_vNodes, net.cpp:2277 started
2021-09-01T11:29:03Z LockContention: cs_vNodes, net.cpp:2277 completed (6μs)
2021-09-01T11:29:04Z LockContention: cs_vNodes, net.cpp:2242 started
2021-09-01T11:29:04Z LockContention: cs_vNodes, net.cpp:2242 completed (3μs)

Co-authored-by: Hennadii Stepanov <32963518+hebasto@users.noreply.github.com>
Co-authored-by: practicalswift <practicalswift@users.noreply.github.com>

jonatack commented:
Moving this out of draft now, ready for review.

laanwj commented:
Code review ACK ccd73de1dd969097d34634c2be2fc32b03fbd09e

MarcoFalke commented on src/sync.h:
```diff
@@ -6,7 +6,10 @@
+#include <logging.h>
 #include <threadsafety.h>
+#include <tinyformat.h>
+#include <util/strencodings.h>
```
why are those includes needed?

jonatack replied:
dropped

MarcoFalke commented on src/sync.h:
```diff
@@ -6,7 +6,10 @@
 #ifndef BITCOIN_SYNC_H
 #define BITCOIN_SYNC_H
 
+#include <logging.h>
```
nit: If `PrintLockContention` stayed in the cpp file, this include can be avoided (it indirectly includes the heavy filesystem header).

jonatack replied:
Good point, though we would have to essentially move `UniqueLock::Enter()` to the cpp file, due to calling `LogAcceptCategory(BCLog::LOCK)`. Punting on this for now.

jonatack replied:
> nit: If `PrintLockContention` stayed in the cpp file, this include can be avoided (it indirectly includes the heavy filesystem header).

agree, dropped the commit changing this

maflcko reviewed (commented):
cr ACK ccd73de1dd969097d34634c2be2fc32b03fbd09e

practicalswift commented:
Concept ACK

Nice idea. I've used something similar in my local tree when analysing lock contention locally.

Non-blocking suggestions:
* In my version I printed the duration too to make it clear from the log entry if the lock contention was significant or not. Perhaps worth adding here too?
* In addition to `__FILE__` and `__LINE__` it would be nice to have `__FUNCTION__` printed as part of the log message too.

To rule out any accidental performance regressions: would it be possible to benchmark before vs after this PR (with and without `-debug=lock`) in some intelligent way?

jonatack commented:
> In my version I printed the duration too to make it clear from the log entry if the lock contention was significant or not. Perhaps worth adding here too?
> 
> In addition to `__FILE__` and `__LINE__` it would be nice to have `__FUNCTION__` printed as part of the log message too.

I think I'd need your help (propose a commit to pull in?) or you could do it as a follow-up.
 
> To rule out any accidental performance regressions: would it be possible to benchmark before vs after this PR (with and without `-debug=lock`) in some intelligent way?

Perhaps!  I'm not sure anyone besides developers would use the lock logging; it's still fairly verbose. In the worst case I suppose we could add a warning in the RPC logging help.

jonatack commented:
@laanwj: following up on our IRC convo in `#bitcoin-core-builds`, as a test just pushed a version with the `HAVE_THREAD_LOCAL` conditional compilation directives removed to see what happens...

jonatack commented:
To-do:
- add a smoke test with -debug=lock
- do some sanity invocation frequency and benchmarking checks
- in a follow-up, try removing `HAVE_THREAD_LOCAL`

MarcoFalke commented:
> add a smoke test with -debug=lock

Shouldn't this be enabled in the functional test already by default?

jonatack commented:
> Shouldn't this be enabled in the functional test already by default?

You're right, TIL.  I was thinking of adding something like this assertion but probably not worth it or reliable enough:

```python3
        with node.assert_debug_log(expected_msgs=["Lock contention: cs_main"]):
            node.generate(COINBASE_MATURITY)
```

theStack commented:
Concept ACK

jonatack commented:
Observations

- `void Mutex::UniqueLock::lock()` acquires the mutex and blocks until it gains access to it

- `bool Mutex::UniqueLock::try_lock()` doesn't block but instead immediately returns whether it acquired the mutex, and it can be used by `lock()` internally as part of the deadlock-avoidance algorithm

- added the following patch and tested a regular, non-debug clang 13 build of bitcoind synced on mainnet
```diff
        if (LogAcceptCategory(BCLog::LOCK)) {
             if (!Base::try_lock()) {
                 LogPrintf("Lock contention: %s, %s:%d\n", pszName, pszFile, nLine);
                 Base::lock();
+            } else {
+                LogPrintf("No lock contention\n");
             }
```
 
- with this, I'm seeing `src/sync.h::UniqueLock::Enter()` invoked 2500-5000 times per second, sampled roughly an hour after startup

- of those, `try_lock()` returns true 99.96% to 99.98% of the time. In the true case the lock logging has essentially the same performance as when it is off, plus the cost of `if (LogAcceptCategory(BCLog::LOCK))` that evaluates to `m_categories.load(std::memory_order_relaxed) & category) != 0` 

- thus 0.02% to 0.04% of the time `try_lock()` returns false, the contention is logged and `lock()` is then called, in which case the additional overhead, if any depending on how long the mutex is blocked, is of trying the lock and initially returning false...which is fairly rare 

- Overall, lock contention seems to be relatively infrequent. The most common one I see is `cs_vNodes` in `net.cpp::2242/2277::ThreadMessageHandler`.

jonatack commented:
Observations

- `void Mutex::UniqueLock::lock()` acquires the mutex and blocks until it gains access to it

- `bool Mutex::UniqueLock::try_lock()` doesn't block but instead immediately returns whether it acquired the mutex, and it can be used by `lock()` internally as part of the deadlock-avoidance algorithm

- added the following patch and tested a regular, non-debug clang 13 build of bitcoind synced on mainnet
```diff
        if (LogAcceptCategory(BCLog::LOCK)) {
             if (!Base::try_lock()) {
                 LogPrintf("Lock contention: %s, %s:%d\n", pszName, pszFile, nLine);
                 Base::lock();
+            } else {
+                LogPrintf("No lock contention\n");
             }
```
 
- with this, I'm seeing `src/sync.h::UniqueLock::Enter()` invoked 2500-5000 times per minute, sampled roughly an hour after startup

- of those, `try_lock()` returns true 99.96% to 99.98% of the time. In the true case the lock logging has essentially the same performance as when it is off, plus the cost of `if (LogAcceptCategory(BCLog::LOCK))` that evaluates to `m_categories.load(std::memory_order_relaxed) & category) != 0` 

- thus 0.02% to 0.04% of the time `try_lock()` returns false, the contention is logged and `lock()` is then called, in which case the additional overhead compared to returning true, if any depending on how long the mutex was blocked, is of trying the lock and initially returning false

- Overall, lock contention seems to be relatively infrequent as a percentage of locks. I'm seeing a few per minute. The most common one I see is `cs_vNodes` in `net.cpp::2242/2277::ThreadMessageHandler`.

theStack reviewed (approved):
Code-review ACK 6e7b6bfee377cdb1ea700af8084b08f79cbe5ee5

hebasto commented:
Concept ACK.

hebasto commented on src/sync.h:
```diff
@@ -138,14 +135,14 @@ class SCOPED_LOCKABLE UniqueLock : public Base
-        if (!Base::try_lock()) {
-            PrintLockContention(pszName, pszFile, nLine);
-#endif
+        if (LogAcceptCategory(BCLog::LOCK)) {
+            if (!Base::try_lock()) {
+                LogPrintf("Lock contention: %s, %s:%d\n", pszName, pszFile, nLine);
+                Base::lock();
+            }
+        } else {
```
```suggestion
        if (LogAcceptCategory(BCLog::LOCK) && !Base::try_lock()) {
            LogPrintf("Lock contention: %s, %s:%d\n", pszName, pszFile, nLine);
        }
```

?

jonatack replied:
Thanks. If I understand the suggestion, this would only lock if the lock logging is on and `try_lock()` succeeds.

hebasto replied:
```cpp
    void Enter(const char* pszName, const char* pszFile, int nLine)
    {
        EnterCritical(pszName, pszFile, nLine, Base::mutex());
        if (LogAcceptCategory(BCLog::LOCK) && !Base::try_lock()) {
            LogPrintf("Lock contention: %s, %s:%d\n", pszName, pszFile, nLine);
        }
        Base::lock();
    }
```

hebasto replied:
The same behavior as with your current code.

hebasto replied:
nm, my suggestion is wrong.

jaysonmald35 replied:
Thank u

hebasto commented:
>     * benchmarking Mutex::UniqueLock `lock()` versus `try_lock()` and the overhead of `LogAcceptCategory(BCLog::LOCK)`, which evaluates to `m_categories.load(std::memory_order_relaxed) & category) != 0`

I think the benchmarking results could depend on platform.

Could we sacrifice the ability to turn "lock" debugging on/off at runtime, and use just a const?

jonatack commented:
> Could we sacrifice the ability to turn "lock" debugging on/off at runtime, and use just a const?

I could be wrong, but it seems these options are low-level, fast, and near constant time compared to system calls, memory/cache coherency and fences, wait queues and (particularly) lock contentions.

jonatack commented:
> Could we sacrifice the ability to turn "lock" debugging on/off at runtime, and use just a const?

I could be wrong, and this goes way beyond my competence, but readings suggest that these options are low-level, fast, and near constant time compared to system calls, memory/cache coherency and fences, wait queues and (particularly) lock contentions.  (Would a difference be apparent in, say, https://bitcoinperf.com? I don't know, but if this helps us reduce contentions, that may be a win. @jamesob, @0xB10C, @martinus thoughts?)

hebasto commented:
> > Could we sacrifice the ability to turn "lock" debugging on/off at runtime, and use just a const?
> 
> I could be wrong, and this goes way beyond my competence, but readings suggest that these options are low-level, fast, and near constant time compared to system calls, memory/cache coherency and fences, wait queues and (particularly) lock contentions. Would a difference be apparent in, say, https://bitcoinperf.com? I don't know, but if this helps us reduce contentions, that may be a win.

And `m_categories.load(std::memory_order_relaxed)` could cause cache lines switching, no?

jonatack commented:
> And `m_categories.load(std::memory_order_relaxed)` could cause cache lines switching, no?

Alternative version:
```cpp
    void Enter(const char* pszName, const char* pszFile, int nLine)
    {
        EnterCritical(pszName, pszFile, nLine, Base::mutex());
        if (Base::try_lock()) return;
        Base::lock();
        if (LogAcceptCategory(BCLog::LOCK)) LogPrintf("Lock contention: %s, %s:%d\n", pszName, pszFile, nLine);
    }
```

hebasto commented:
> > And `m_categories.load(std::memory_order_relaxed)` could cause cache lines switching, no?
> 
> Alternative version:
> 
> ```c++
>     void Enter(const char* pszName, const char* pszFile, int nLine)
>     {
>         EnterCritical(pszName, pszFile, nLine, Base::mutex());
>         if (Base::try_lock()) return;
>         Base::lock();
>         LogPrint(BCLog::LOCK, "Lock contention: %s, %s:%d\n", pszName, pszFile, nLine);
>     }
> ```

Nice!

Maybe put logging before the blocking call `Base::lock();` as it is now?

jonatack commented:
> And `m_categories.load(std::memory_order_relaxed)` could cause cache lines switching, no?

I don't know.

> Maybe put logging before the blocking call `Base::lock();` as it is now? The `LogPrint` call takes some time, and it could increase probability of releasing a mutex by another thread before the `Base::lock()` call.

Seems like a good idea--done.

hebasto reviewed (approved):
ACK df149f89dcd0c457c61d463e36c057bb064783b8, tested on Linux Mint 20.2 (x86_64).

> > In my version I printed the duration too to make it clear from the log entry if the lock contention was significant or not. Perhaps worth adding here too?
> 
> I think I'd need your help (propose a commit to pull in?) or you could do it as a follow-up.

```diff
--- a/src/sync.h
+++ b/src/sync.h
@@ -7,6 +7,7 @@
 #define BITCOIN_SYNC_H
 
 #include <logging.h>
+#include <logging/timer.h>
 #include <threadsafety.h>
 #include <util/macros.h>
 
@@ -136,7 +137,7 @@ private:
     {
         EnterCritical(pszName, pszFile, nLine, Base::mutex());
         if (Base::try_lock()) return;
-        LogPrint(BCLog::LOCK, "Lock contention: %s, %s:%d\n", pszName, pszFile, nLine);
+        LOG_TIME_MILLIS_WITH_CATEGORY(strprintf("Lock contention: %s, %s:%d", pszName, pszFile, nLine), BCLog::LOCK);
         Base::lock();
     }
 
```

as a follow up?

jonatack commented:
> as a follow up?

Thanks--was about to re-push to drop the third commit. Will try your suggestion before re-pushing.

hebasto commented:
> > as a follow up?
> 
> Thanks--was about to re-push to drop the third commit. Will try your suggestion before re-pushing.

My only concerns are about wording -- "Enter: Lock contention: cs_vNodes, net.cpp:2242 completed (0.07ms)" -- "contention completed"?

jonatack commented:
> My only concerns are about wording -- "Enter: Lock contention: cs_vNodes, net.cpp:2242 completed (0.07ms)" -- "contention completed"?

I'm seeing the following (all of the times are 0.00ms or 0.01ms so far)
```
2021-09-01T11:03:59Z [msghand] PrintLockContention: Lock contention: cs_vNodes, net.cpp:2242 started
2021-09-01T11:03:59Z [msghand] PrintLockContention: Lock contention: cs_vNodes, net.cpp:2242 completed (0.01ms)
```

`timer.h::LOG_TIME_MILLIS_WITH_CATEGORY` times the execution of the whole Enter() function, the placement of `Base::lock();` doesn't matter for it?

jaysonmald35 commented:
Can I pay u

jonatack commented:
<details><summary>How about microseconds?</summary><p>

```diff
diff --git a/src/logging/timer.h b/src/logging/timer.h
index 159920e397..cadfe395ee 100644
--- a/src/logging/timer.h
+++ b/src/logging/timer.h
@@ -93,6 +93,8 @@ private:
 } // namespace BCLog
 
 
+#define LOG_TIME_MICROS_WITH_CATEGORY(end_msg, log_category) \
+    BCLog::Timer<std::chrono::microseconds> PASTE2(logging_timer, __COUNTER__)(__func__, end_msg, log_category)
 #define LOG_TIME_MILLIS_WITH_CATEGORY(end_msg, log_category) \
     BCLog::Timer<std::chrono::milliseconds> PASTE2(logging_timer, __COUNTER__)(__func__, end_msg, log_category)
 #define LOG_TIME_SECONDS(end_msg) \
diff --git a/src/sync.cpp b/src/sync.cpp
index a7b63ea986..eace86d9dd 100644
--- a/src/sync.cpp
+++ b/src/sync.cpp
@@ -24,9 +24,9 @@
 #include <utility>
 #include <vector>
 
-void PrintLockContention(const char* pszName, const char* pszFile, int nLine)
+void LockContention(const char* pszName, const char* pszFile, int nLine)
 {
-    LOG_TIME_MILLIS_WITH_CATEGORY(strprintf("%s, %s:%d", pszName, pszFile, nLine), BCLog::LOCK);
+    LOG_TIME_MICROS_WITH_CATEGORY(strprintf("%s, %s:%d", pszName, pszFile, nLine), BCLog::LOCK);
 }
 
 #ifdef DEBUG_LOCKORDER
diff --git a/src/sync.h b/src/sync.h
index 252dc95845..0a8ec9e58f 100644
--- a/src/sync.h
+++ b/src/sync.h
@@ -126,7 +126,8 @@ using RecursiveMutex = AnnotatedMixin<std::recursive_mutex>;
 /** Wrapped mutex: supports waiting but not recursive locking */
 typedef AnnotatedMixin<std::mutex> Mutex;
 
-void PrintLockContention(const char* pszName, const char* pszFile, int nLine);
+/** Print a lock contention to the log */
+void LockContention(const char* pszName, const char* pszFile, int nLine);
 
 /** Wrapper around std::unique_lock style lock for Mutex. */
 template <typename Mutex, typename Base = typename Mutex::UniqueLock>
@@ -137,7 +138,7 @@ private:
     {
         EnterCritical(pszName, pszFile, nLine, Base::mutex());
         if (Base::try_lock()) return;
-        PrintLockContention(pszName, pszFile, nLine);
+        LockContention(pszName, pszFile, nLine);
         Base::lock();
     }
```


```
2021-09-01T11:29:03Z [net] LockContention: cs_vNodes, net.cpp:1251 started
2021-09-01T11:29:03Z [net] LockContention: cs_vNodes, net.cpp:1251 completed (5.00μs)
2021-09-01T11:29:03Z [net] LockContention: pnode->cs_vSend, net.cpp:1373 started
2021-09-01T11:29:03Z [net] LockContention: pnode->cs_vSend, net.cpp:1373 completed (31.00μs)
2021-09-01T11:29:03Z [msghand] LockContention: cs_vNodes, net.cpp:2277 started
2021-09-01T11:29:03Z [msghand] LockContention: cs_vNodes, net.cpp:2277 completed (6.00μs)
2021-09-01T11:29:04Z [msghand] LockContention: cs_vNodes, net.cpp:2242 started
2021-09-01T11:29:04Z [msghand] LockContention: cs_vNodes, net.cpp:2242 completed (3.00μs)
```
</p></details>

martinus commented:
> And `m_categories.load(std::memory_order_relaxed)` could cause cache lines switching, no?

yes, the only guarantee with `memory_order_relaxed` is that the operation itself is atomic, but other operations are allowed to be  shuffled that by the compiler. I personally try to avoid using any of the memory order stuff. It's too easy to get wrong. 

By the way, I think that in that patch: 

```patch
        if (LogAcceptCategory(BCLog::LOCK)) {
             if (!Base::try_lock()) {
                 LogPrintf("Lock contention: %s, %s:%d\n", pszName, pszFile, nLine);
                 Base::lock();
+            } else {
+                LogPrintf("No lock contention\n");
             }
```

If `Base::try_lock()` fails strictly speaking that does not necessarily mean that there's a lock contention. `std::mutex`'s `try_lock()` can also spuriously fail even when no other thread is currently holding a lock: https://en.cppreference.com/w/cpp/thread/mutex/try_lock I have no idea though how often that can happen.

jonatack committed:
log, timer: add timing macro in usec LOG_TIME_MICROS_WITH_CATEGORY

and update BCLog::LogMsg() to omit irrelevant decimals for microseconds
and skip unneeded code and math.

jonatack commented:
Thanks for the ideas and feedback!  Updated with time duration logging in microseconds like in https://github.com/bitcoin/bitcoin/pull/22736#issuecomment-910198774.

> If `Base::try_lock()` fails strictly speaking that does not necessarily mean that there's a lock contention. `std::mutex`'s `try_lock()` can also spuriously fail even when no other thread is currently holding a lock: [en.cppreference.com/w/cpp/thread/mutex/try_lock](https://en.cppreference.com/w/cpp/thread/mutex/try_lock) I have no idea though how often that can happen.

Good point; I read this before and admit to discounting it a little in my mind since we don't know the frequency.  It seems like a good reason to log the contention duration.

jonatack commented:
Thanks for the ideas and feedback. Updates:

- added time duration in microseconds to the lock contention logging
- improved `BCLog::LogMsg()` to omit irrelevant decimals for microseconds and skip unnecessary code/math

> If `Base::try_lock()` fails strictly speaking that does not necessarily mean that there's a lock contention. `std::mutex`'s `try_lock()` can also spuriously fail even when no other thread is currently holding a lock: [en.cppreference.com/w/cpp/thread/mutex/try_lock](https://en.cppreference.com/w/cpp/thread/mutex/try_lock). I have no idea though how often that can happen.

Good point. I read this before and admit to discounting it a little since we don't know the frequency.  It seems to be an additional reason to log the contention duration.

hebasto reviewed (approved):
re-ACK 7e698732836121912f179b7c743a72dd6fdffa72, added a contention duration to the log message since my [previous](https://github.com/bitcoin/bitcoin/pull/22736#pullrequestreview-743764606) review.

theStack reviewed (approved):
re-ACK 7e698732836121912f179b7c743a72dd6fdffa72 🔏 ⏲️ 

Since my last ACK, the lock contention logging was improved by also showing the duration in micro-seconds, and an alternate (more elegant) version of `UniqueLock::Enter` is now used that doesn't need to check whether the `lock` category is on if there is no contention (https://github.com/bitcoin/bitcoin/pull/22736#issuecomment-910110140).

Tested by building with clang 10.0.1 on OpenBSD 6.9 and running `./src/bitcoind -signet -debug=lock` on an empty signet datadir.

On my system, during IBD the vast majority of lock contentions happen in
* cs_main, net_processing.cpp:1476
* m_mutex, ./checkqueue.h:168
* m_mutex, ./checkqueue.h:77

After IBD is finished, _all_ lock contentions are in
* cs_vNodes, net.cpp (different lines)

Most contentions are in the range of 5-10us. (Note that these results are not based on deeper analysis, but on what I saw scrolling through the logs after a few minutes).

jonatack commented:
Thanks for testing! For me, there seems to be a higher frequency of contention and occuring in more places on mainnet than on signet, presumably as the scale/numbers are larger. Also, a big difference in contention durations between a regular build and a debug build, which makes sense.

MarcoFalke commented on src/logging/timer.h:
```diff
@@ -58,12 +58,14 @@ class Timer
-        } else if (std::is_same<TimeType, std::chrono::milliseconds>::value) {
+        if (std::is_same<TimeType, std::chrono::milliseconds>::value) {
             units = "ms";
             divisor = 1000.;
```
unrelated: Could make sense to inline those as well to avoid the heterogeneous code flow?

jonatack replied:
good idea, done (and more) in #22904

martinus commented on src/sync.cpp:
```diff
@@ -27,10 +28,9 @@
 {
-    LogPrint(BCLog::LOCK, "LOCKCONTENTION: %s\n", pszName);
-    LogPrint(BCLog::LOCK, "Locker: %s:%d\n", pszFile, nLine);
+    LOG_TIME_MICROS_WITH_CATEGORY(strprintf("%s, %s:%d", pszName, pszFile, nLine), BCLog::LOCK);
```
I think `LockContention` doesn't do what it is supposed to do. At least it doesn't log the time that the contention actually takes: the macro creates an `BCLog::Timer` instance where constructor measures start time, and destructor prints finished. But since all that is done inside the `LockContention` function, this happens instantly and doesn't measure the actual lock contention.

I've verified that in https://github.com/bitcoin/bitcoin/pull/21006 because I wanted to see how long the contentions are, but it always prints 0µs or 1µs. When I move the macro into `sync.h` and replace 

```cpp
LockContention(pszName, pszFile, nLine); // log the contention`
```
with
```cpp
LOG_TIME_MICROS_WITH_CATEGORY(strprintf("%s, %s:%d", pszName, pszFile, nLine), BCLog::LOCK);
```
I get reasonable contention times, like so:

```
2021-09-06T17:18:23Z Enter: cs_main, rest.cpp:281 completed (65406μs)
2021-09-06T17:18:23Z Enter: cs_main, rest.cpp:281 completed (65331μs)
2021-09-06T17:18:23Z Enter: cs_main, rest.cpp:281 completed (65074μs)
2021-09-06T17:18:23Z Enter: cs_main, rest.cpp:281 completed (65103μs)
2021-09-06T17:18:23Z Enter: cs_main, net_processing.cpp:4051 completed (67687μs)
```

jonatack replied:
Interesting. Right now on mainnet with a non-debug clang 13 build running since 6 hours (2 cores x 2.50GHz), I see varying durations of 1-50 usec, most often 2-30.  I suspect your machine is much faster than mine. Will try your suggestion.

jonatack replied:
@martinus This does seem better, now seeing durations of 1 to several million usec (mainnet, non-debug build). Thanks for reporting. Proposed fix in #22904.
